src/Storages/MaterializedView/RefreshTask.cpp

```diff
@@ -216,28 +216,28 @@ void RefreshTask::checkAlterIsPossible(const DB::ASTRefreshStrategy & new_strate
 
 void RefreshTask::alterRefreshParams(const DB::ASTRefreshStrategy & new_strategy)
 {
-    {
-        std::lock_guard guard(mutex);
+    std::lock_guard guard(mutex);
 
-        refresh_schedule = RefreshSchedule(new_strategy);
-        std::vector<StorageID> deps;
-        if (new_strategy.dependencies)
-            for (auto && dependency : new_strategy.dependencies->children)
-                deps.emplace_back(dependency->as<const ASTTableIdentifier &>());
+    refresh_schedule = RefreshSchedule(new_strategy);
+    std::vector<StorageID> deps;
+    if (new_strategy.dependencies)
+        for (auto && dependency : new_strategy.dependencies->children)
+            deps.emplace_back(dependency->as<const ASTTableIdentifier &>());
 
-        /// Update dependency graph.
-        if (set_handle)
-            set_handle.changeDependencies(deps);
+    /// Update dependency graph.
+    if (set_handle)
+        set_handle.changeDependencies(deps);
 
-        scheduleRefresh(guard);
-        scheduling.dependencies_satisfied_until = std::chrono::sys_seconds(std::chrono::seconds(-1));
+    scheduleRefresh(guard);
+    scheduling.dependencies_satisfied_until = std::chrono::sys_seconds(std::chrono::seconds(-1));
+
+    refresh_settings = {};
+    if (new_strategy.settings != nullptr)
+        refresh_settings.applyChanges(new_strategy.settings->changes);
 
-        refresh_settings = {};
-        if (new_strategy.settings != nullptr)
-            refresh_settings.applyChanges(new_strategy.settings->changes);
-    }
     /// In case refresh period changed.
-    view->getContext()->getRefreshSet().notifyDependents(view->getStorageID());
+    if (view)
+        view->getContext()->getRefreshSet().notifyDependents(view->getStorageID());
```

Comment:
Are you sure it's ok to lock RefreshSet::mutex while holding RefreshTask::mutex, i.e. that they're never locked in the opposite order? It seems so, but it also seems easy to avoid relying on it here: call `getRefreshSet()` and `getStorageID()` with mutex locked, then unlock and call `notifyDependents`.

Reply:
Ah, the problem is not RefreshSet, it's other RefreshTask mutexes (through notifyDependents -> RefreshTask::notify). It's maybe possible to ALTER two tables to depend on each other, then have them notify each other at the same time and deadlock trying to lock the two mutexes in different order. So please move the call out of the lock.

Reply:
Good point. Indeed, I came to the same conclusion that lock-order-inversion may happen. Fixed.

Note: I have to use global context, since `getRefreshSet` returns a reference, and RefreshSet is not copyable (and should not), but this should be safe, since it is anyway shared.